source/base/polynomials_rt_bubbles.cc

```diff
@@ -0,0 +1,504 @@
+  monomials(k+2),
+  n_pols(compute_n_pols(k))
+{
+  Assert (dim >= 2, ExcImpossibleInDim(dim));
+
+  for (unsigned int i=0; i<monomials.size(); ++i)
+    monomials[i] = Polynomials::Monomial<double> (i);
+}
+
+
+template <int dim>
+void
+PolynomialsRT_Bubbles<dim>::compute (const Point<dim>         &unit_point,
+                                     std::vector<Tensor<1,dim> > &values,
+                                     std::vector<Tensor<2,dim> > &grads,
+                                     std::vector<Tensor<3,dim> > &grad_grads,
+                                     std::vector<Tensor<4,dim> > &third_derivatives,
+                                     std::vector<Tensor<5,dim> > &fourth_derivatives) const
+{
+  Assert(values.size()==n_pols || values.size()==0,
+         ExcDimensionMismatch(values.size(), n_pols));
+  Assert(grads.size()==n_pols || grads.size()==0,
+         ExcDimensionMismatch(grads.size(), n_pols));
+  Assert(grad_grads.size()==n_pols || grad_grads.size()==0,
+         ExcDimensionMismatch(grad_grads.size(), n_pols));
+  Assert(third_derivatives.size()==n_pols || third_derivatives.size()==0,
+         ExcDimensionMismatch(third_derivatives.size(), n_pols));
+  Assert(fourth_derivatives.size()==n_pols || fourth_derivatives.size()==0,
+         ExcDimensionMismatch(fourth_derivatives.size(), n_pols));
+
+  // Third and fourth derivatives are not implemented
+  (void)third_derivatives;
+  Assert(third_derivatives.size()==0,
+         ExcNotImplemented());
+  (void)fourth_derivatives;
+  Assert(fourth_derivatives.size()==0,
+         ExcNotImplemented());
+
+  const unsigned int n_sub = raviart_thomas_space.n();
+
+  // Guard access to the scratch arrays in the following block
+  // using a mutex to make sure they are not used by multiple threads
+  // at once
+
+  static Threads::Mutex mutex;
+  Threads::Mutex::ScopedLock lock(mutex);
+
+  static std::vector<Tensor<1,dim> > p_values;
+  static std::vector<Tensor<2,dim> > p_grads;
+  static std::vector<Tensor<3,dim> > p_grad_grads;
+  static std::vector<Tensor<4,dim> > p_third_derivatives;
+  static std::vector<Tensor<5,dim> > p_fourth_derivatives;
+
+  p_values.resize((values.size() == 0) ? 0 : n_sub);
+  p_grads.resize((grads.size() == 0) ? 0 : n_sub);
+  p_grad_grads.resize((grad_grads.size() == 0) ? 0 : n_sub);
+
+  // This is the Raviart-Thomas part of the space
+  raviart_thomas_space.compute (unit_point, p_values, p_grads, p_grad_grads, p_third_derivatives, p_fourth_derivatives);
+  for (unsigned int i=0; i<p_values.size(); ++i)
+    values[i] = p_values[i];
+  for (unsigned int i=0; i<p_grads.size(); ++i)
+    grads[i] = p_grads[i];
+  for (unsigned int i=0; i<p_grad_grads.size(); ++i)
+    grad_grads[i] = p_grad_grads[i];
+
+  // Next we compute the polynomials and derivatives
+  // of the curl part of the space
+  std::vector<std::vector<double> > monoval_plus(dim, std::vector<double>(4));
+  std::vector<std::vector<double> > monoval(dim, std::vector<double>(4));
```

Comment:
You could avoid the repeated memory allocation here if you made these variables `static` and then used a mutex/lock pair as above to guard access to them.

But since you know the sizes of the objects, the simplest approach may actually be to declare them as statically-sized arrays:
```
  double monoval_plus[dim][4];
```
and similarly for the other variables.